### DISCUSSION TITLE
Rules and IF statements

### DISCUSSION THREAD

**Discussion by tsunfinger** [arendst/Tasmota#20756] [Q&A] [answered]
using Tasmota 13.3.0 on shelly and playing with rules. Can somebody give a hint please, why this rule isn't working as expected?
single command on this rule is working (only "var5 55" instead of "var5 55; var6 22")

rule2
on var2#state do
backlog0 IF (%var4% == 1) var5 55; var6 22 ENDIF; var7 11
endon

var4 2
18:02:37.905 RUL: VAR2#STATE performs "backlog0 IF (2 == 1) var5 55; var6 22 ENDIF; var7 11"
18:02:37.960 MQT: stat/tasmota_4F9978/RESULT = {"If":"Done"}
18:02:37.976 MQT: stat/tasmota_4F9978/RESULT = {"Var6":"22 ENDIF"}
18:02:37.990 MQT: stat/tasmota_4F9978/RESULT = {"Var7":"11"}

var4 1
18:04:12.134 RUL: VAR2#STATE performs "backlog0 IF (1 == 1) var5 55; var6 22 ENDIF; var7 11"
18:04:12.198 MQT: stat/tasmota_4F9978/RESULT = {"If":"Done"}
18:04:12.210 MQT: stat/tasmota_4F9978/RESULT = {"Var6":"22 ENDIF"}
18:04:12.223 MQT: stat/tasmota_4F9978/RESULT = {"Var7":"11"}

**Comment by barbudor** [comment]
Remove 'backlog0' as not required and probably the reason for your problems

**Reply by tsunfinger** [reply]
ok, but in this case var7 won't be set
any idea for a workaround? set var7 twice - inside IF statement and after with ELSEIF?

**Reply by barbudor** [reply, marked as answer]
As suggested in docs
![Screenshot_2024-02-18-18-50-33-088_com android chrome-edit](https://github.com/arendst/Tasmota/assets/2996042/85c6c586-5723-43e6-9e19-99273e8f1924)

**Reply by sfromis** [reply]
The Backlog command does not play very well with IF/ENDIF. Seeing `{"Var6":"22 ENDIF"}` means that the parsing broke down, and the Var command was outside the parsing for ENDIF to match the IF.

If you insert an extra `;` the parsing of IF/ENDIF is no longer broken. This could be any other command, but no command is just an empty command, and therefore can work.
```
rule2
on var2#state do
backlog0 ; IF (%var4% == 1) var5 55; var6 22 ENDIF; var7 11
endon
```
I'm not sure if this is a generally robust solution, but it worked in this specific case at least.

**Reply by tsunfinger** [reply]
this workaround seems to be running. I built something with ELSEIF but doesn't seems to be running as well
`rule2
on var1#state do
IF (%value%  ==  1) var5 55; var7 77 ELSEIF var7 77 ENDIF
endon`

08:18:57.922 CMD: var1 2
08:18:57.927 MQT: stat/tasmota_4F9978/RESULT = {"Var1":"2"}
08:18:57.949 RUL: VAR1#STATE performs "IF (2  ==  1) var5 55; var7 77 ELSEIF var7 77 ENDIF"
08:18:57.953 MQT: stat/tasmota_4F9978/RESULT = {"If":"Done"}
08:19:12.515 CMD: var1 1
08:19:12.519 MQT: stat/tasmota_4F9978/RESULT = {"Var1":"1"}
08:19:12.545 RUL: VAR1#STATE performs "IF (1  ==  1) var5 55; var7 77 ELSEIF var7 77 ENDIF"
08:19:12.550 MQT: stat/tasmota_4F9978/RESULT = {"If":"Done"}
08:19:12.768 MQT: stat/tasmota_4F9978/RESULT = {"Var5":"55"}
08:19:13.018 MQT: stat/tasmota_4F9978/RESULT = {"Var7":"77"}

**Reply by sfromis** [reply]
To be valid syntax, that should be `ELSE` instead of `ELSEIF`, as you have no new condition.

**Reply by tsunfinger** [reply]
Ah sry, stupid copy&paste error from my site. Yes
`rule2 on var1#state do IF (%value%  ==  1) var5 55; var7 77 ELSE var7 77 ENDIF endon`
is working as expected
